### PR TITLE
Close #584 Memtest Without External Tools

### LINKED ISSUE #584: Memtest: External Tools
`cuda-memtest` has some unnecessary external command dependencies such as [gawk and nvidia-smi](https://github.com/ax3l/picongpu/commit/a9cf5bec03af068d41d966a83b09464c9e67eb1f#commitcomment-8361308) to get the serial number of the device.

### PR DESCRIPTION
**Pull request by ax3l**
Close #584

- removes the "nvidia-smi", "grep" and "gawk" calls
- adds optional support for `libnvidia-ml` ([NVML](https://developer.nvidia.com/nvidia-management-library-nvml)) which is either part of the driver or the `gdk` ([GPU Deployment Kit](https://developer.nvidia.com/gpu-deployment-kit)) to:
  - measure the device temperature
  - get the device serial number (by `index`, let's hope it is the same index as in cuda for now - else we have to use the [PIC bus id](http://docs.nvidia.com/deploy/nvml-api/group__nvmlDeviceQueries.html#group__nvmlDeviceQueries_1gcfea8661435e6a8f26485c5c42eadd2f) if that works for 2-chip cards like K80)
- adds initial support for 8 GPUs/node (previous config was 4 GPUs) but needs additional testing
- thanks to @jirikraus for the support and initial hint to use NVML

### CODE REVIEW

File: src/cuda_memtest/cuda_memtest.h [hunk 1]
```diff
@@ -130,6 +138,19 @@ extern void get_driver_info(char* info, unsigned int len);
...
+/**
+ * Captures NVML errors and prints messages to stdout, including line number and file.
+ *
+ * @param cmd command with nvmlReturn_t return value to check
+ */
+#define NVML_CHECK(cmd) {nvmlReturn_t returnVal = cmd; if(returnVal!=NVML_SUCCESS){std::cerr<<"<"<<__FILE__<<">:"<<__LINE__<<std::endl; throw std::runtime_error(std::string("[NVML] Error: ") + std::string(nvmlErrorString(returnVal)));}}
+
+#define NVML_CHECK_MSG(cmd,msg) {nvmlReturn_t returnVal = cmd; if(returnVal!=NVML_SUCCESS){std::cerr<<"<"<<__FILE__<<">:"<<__LINE__<<msg<<std::endl; throw std::runtime_error(std::string("[NVML] Error: ") + std::string(nvmlErrorString(returnVal)));}}
+
+#define NVML_CHECK_NO_EXCEP(cmd) {nvmlReturn_t returnVal = cmd; if(returnVal!=NVML_SUCCESS){printf("[NVML] Error: <%s>:%i ",__FILE__,__LINE__);}}
```
Comment (psychocoderHPC) [inline src/cuda_memtest/cuda_memtest.h]: Please use `std::err` instead of `printf`.

File: src/cuda_memtest/tests.cu [hunk 2]
```diff
@@ -130,8 +129,14 @@ error_checking(const char* msg, unsigned int blockidx)
 	FPRINTF("ERROR: %s",  driver_info);
 	emsg += sprintf(emsg, "ERROR: %s", driver_info);
 
-	FPRINTF("ERROR: The unit serial number is %llu\n",  serial_number);
-	emsg += sprintf(emsg, "ERROR: The unit serial number is %llu\n",  serial_number);
+#if !defined(NVML_DEVICE_SERIAL_BUFFER_SIZE)
+    char devSerialNum[] = "unknown (no NVML found)";
+#else
+    char devSerialNum[NVML_DEVICE_SERIAL_BUFFER_SIZE];
```
Comment (psychocoderHPC) [inline src/cuda_memtest/tests.cu]: please fix the line alignment

**Comment by PrometheusPi** [comment]
:+1: great job

**Comment by psychocoderHPC** [comment]
+1 and big thanks

**Comment by ax3l** [comment]
thx guys, np :)

**Pull request merged**